File: services/analytics-python/app/analytics/service_token.py
"""Mint analytics-python service tokens accepted by core-java's verifier.

The wire format mirrors core-java exactly: a compact HS256 JWS
(``base64url(header).base64url(payload).base64url(HMAC-SHA256)``) with the
issuer ``analytics-python``.
"""

from __future__ import annotations

import time

from app.analytics.jws import encode_segment, sign

ISSUER = "analytics-python"
AUDIENCE = "core-java"


def sign_service_token(
    secret: str, operator_id: str, role: str = "OPERATOR_STAFF", ttl_secs: int = 300
) -> str:
    header = encode_segment({"alg": "HS256", "typ": "JWT"})
    payload = encode_segment(
        {
            "iss": ISSUER,
            "aud": AUDIENCE,
            "typ": "service",
            "exp": int(time.time()) + ttl_secs,
            "operatorId": operator_id,
            "role": role,
        }
    )
    signing_input = f"{header}.{payload}"
    return f"{signing_input}.{sign(signing_input, secret)}"

File: services/gateway-node/src/auth/serviceToken.ts
import jwt from "jsonwebtoken";

import { type Config } from "../config.js";
import { type Identity } from "./identity.js";

export type ServiceTarget = "core-java" | "analytics-python";

export function mintServiceToken(
  identity: Identity,
  target: ServiceTarget,
  config: Pick<
    Config,
    "GATEWAY_SERVICE_TOKEN_SECRET" | "GATEWAY_ANALYTICS_SERVICE_TOKEN_SECRET" | "SERVICE_TOKEN_TTL_SECS"
  >,
): string {
  const claims = {
    iss: "gateway-node",
    aud: target,
    typ: "service",
    sub: identity.sub,
    operatorId: identity.operatorId,
    role: identity.role,
    ...(identity.subscriberId ? { subscriberId: identity.subscriberId } : {}),
  };
  return jwt.sign(claims, secretForTarget(target, config), {
    algorithm: "HS256",
    expiresIn: config.SERVICE_TOKEN_TTL_SECS,
  });
}

function secretForTarget(
  target: ServiceTarget,
  config: Pick<Config, "GATEWAY_SERVICE_TOKEN_SECRET" | "GATEWAY_ANALYTICS_SERVICE_TOKEN_SECRET">,
): string {
  return target === "analytics-python"
    ? config.GATEWAY_ANALYTICS_SERVICE_TOKEN_SECRET
    : config.GATEWAY_SERVICE_TOKEN_SECRET;
}
